modules/unsupported/vector-mosaic/src/main/java/org/geotools/vectormosaic/VectorMosaicFeatureReader.java

```diff
@@ -0,0 +1,218 @@
+                granuleDataStore.dispose();
+            }
+
+            delegateFeature = (SimpleFeature) delegateIterator.next();
+            VectorMosaicGranule granule = VectorMosaicGranule.fromDelegateFeature(delegateFeature);
+            source.initGranule(granule, false);
+            Filter granuleFilter =
+                    source.getSplitFilter(
+                            query, granule.getDataStore(), granule.getGranuleTypeName(), false);
+            granuleDataStore = granule.getDataStore();
+            granuleIterator =
+                    granuleDataStore
+                            .getFeatureSource(granule.getGranuleTypeName())
+                            .getFeatures(granuleFilter)
+                            .features();
+
+            if (granuleIterator.hasNext()) {
+                peekGranule = (SimpleFeature) granuleIterator.next();
+                Feature peek = mergeGranuleAndDelegate(peekGranule, delegateFeature);
+                if (query.getFilter().evaluate(peek)) {
+                    nextGranule = peek;
+                    return true;
+                }
+            }
+        }
+
+        // no more
+        if (granuleIterator != null) {
+            // close the last iterator
+            granuleIterator.close();
+        }
+        if (granuleDataStore != null) {
+            granuleDataStore.dispose();
+        }
+        if (delegateIterator != null) {
+            delegateIterator.close();
+        }
+        return false;
+    }
+
+    /**
+     * Merge the granule feature with the delegate feature
+     *
+     * @param peekGranule the granule feature
+     * @param delegateFeature the delegate feature
+     * @return the merged feature
+     */
+    private Feature mergeGranuleAndDelegate(
+            SimpleFeature peekGranule, SimpleFeature delegateFeature) {
+        SimpleFeatureType mergedType = getFeatureType();
+        List<Object> attributes = new ArrayList<>();
+        peekGranule.getType().getAttributeDescriptors().stream()
+                .forEach(
+                        attributeDescriptor -> {
+                            attributes.add(peekGranule.getAttribute(attributeDescriptor.getName()));
+                        }); // type starts with of all granule attributes
+        delegateFeature.getType().getAttributeDescriptors().stream()
+                .filter(source::isNotMandatoryIndexType)
+                .forEach(
+                        attributeDescriptor -> {
+                            attributes.add(
+                                    delegateFeature.getAttribute(attributeDescriptor.getName()));
+                        }); // then add all delegate attributes except mandatory index types
+
+        return SimpleFeatureBuilder.build(
+                mergedType, attributes, peekGranule.getIdentifier().getID());
+    }
+
+    @Override
+    public void close() throws IOException {
+        if (delegateIterator != null) {
```

Comment:
If the iterator is closed before reaching the end, I believe the delegate store won't be closed.

Reply:
Fixed